gh-111389: expose _PyHASH_INF/BITS/MODULUS/IMAG macros as public

skirpichev opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-111389 -->
* Issue: gh-111389
<!-- /gh-issue-number -->

skirpichev committed:
gh-111389: restoring _PyHASH_INF/BITS/MODULUS/IMAG macroses as public

This partially reverts #107026.

skirpichev committed:
Mention public macros in docs

vstinner commented on Doc/library/sys.rst:
```diff
@@ -1022,19 +1022,22 @@ always available.
 
-      The prime modulus P used for numeric hash scheme
+      The prime modulus P used for numeric hash scheme.  Available also as
+      ``_PyHASH_MODULUS`` macro.
```
Usually we don't add references to the C API in the Python documentation. Moreover, usually we don't document the private API.

skirpichev replied:
Redo in Doc/c-api/hashing.rst per [suggestion](https://github.com/python/cpython/issues/111389#issuecomment-1784940869) in the issue thread.

skirpichev committed:
Revert "Mention public macros in docs"

This reverts commit 2c6002f521bb78d11340b267104b9a64eaa2096f.

skirpichev committed:
Add Doc/c-api/hashing.rst for hash-related C macros

vstinner reviewed (changes requested):
Please don't document private APIs.

skirpichev committed:
Expose macroses as unstable API

vstinner commented on Include/pyhash.h:
```diff
@@ -16,6 +16,27 @@ typedef struct {
+#define _PyHASH_INF 314159
+#define PyUnstable_PyHASH_INF _PyHASH_INF
+#define _PyHASH_IMAG _PyHASH_MULTIPLIER
+#define PyUnstable_PyHASH_IMAG _PyHASH_IMAG
```
Your change adds these API to the limited C API. I don't think that we want that.

I just created Include/cpython/pyhash.h to better distinguish the limited C API from the non-limited C API.

skirpichev replied:
> Your change adds these API to the limited C API. I don't think that we want that.

In past hashing of numeric types was changed in minor releases (3.x.0).  Are there reasons we don't want this freedom in future?

> I suggest waiting until this PR is merged before considering exposing these constants to the public C API.

Sound reasonable.

vstinner replied:
> In past hashing of numeric types was changed in minor releases (3.x.0). Are there reasons we don't want this freedom in future?

It was proposed to create a C API Working Group. The Steering Council [has to take a decision](https://github.com/python/steering-council/issues/210). For now, no new API should be added to the limited C API.

See https://github.com/capi-workgroup/api-evolution/issues/41 and https://github.com/capi-workgroup/api-evolution/issues/42

skirpichev replied:
Ok, added as public.  Docs moved to the hash.rst, news entry also added.  I don't think this could introduce merge conflicts with https://github.com/python/cpython/pull/112095

vstinner reviewed (commented):
I'm working on PR https://github.com/python/cpython/pull/112095 which adds an "infrastructure" to test the PyHash API. I suggest waiting until this PR is merged before considering exposing these constants to the public C API.

skirpichev committed:
Merge branch 'main' into restore-pyhash-and-pylongnew

skirpichev committed:
Merge branch 'main' into restore-pyhash-and-pylongnew

skirpichev committed:
Expose macros as public and move docs to Doc/c-api/hash.rst

skirpichev committed:
news entry

skirpichev committed:
restore link to hashing algorithms

skirpichev committed:
Merge branch 'main' into restore-pyhash-and-pylongnew

vstinner commented on Misc/NEWS.d/next/C API/2023-11-15-09-24-51.gh-issue-111418.FYYetY.rst:
```diff
@@ -0,0 +1,2 @@
+Add :c:macro:`PyHASH_MODULUS`, :c:macro:`PyHASH_BITS`, :c:macro:`PyHASH_INF`
+and :c:macro:`PyHASH_IMAG` C macroses.  Patch by Sergey B Kirpichev.
```
```suggestion
and :c:macro:`PyHASH_IMAG` C macros.  Patch by Sergey B Kirpichev.
```

vstinner commented on Doc/c-api/hash.rst:
```diff
@@ -17,6 +18,21 @@ See also the :c:member:`PyTypeObject.tp_hash` member.
 
+.. c:macro:: PyHASH_MODULUS
+
+   The Mersenne prime ``P = 2**n -1``, used for numeric hash scheme.
```
Please add `.. versionadded:: 3.13` to the added macros.

skirpichev committed:
Update Misc/NEWS.d/next/C API/2023-11-15-09-24-51.gh-issue-111418.FYYetY.rst

Co-authored-by: Victor Stinner <vstinner@python.org>

skirpichev committed:
address review: versionadded

skirpichev committed:
Move to Include/cpython/pyhash.h

skirpichev committed:
docs nitpick

Was rendered: "See also the PyTypeObject.tp_hash member and Hashing
of numeric types for more details about hashing of numeric types."

vstinner commented on Include/cpython/pyhash.h:
```diff
@@ -15,9 +15,14 @@
 #  define _PyHASH_BITS 31
 #endif
 
+#define PyHASH_BITS _PyHASH_BITS
```
I would prefer to do the opposite: keep the old name as a deprecated alias, but define the public name to the value: 3 lines above. Same remark for other constants.

skirpichev replied:
Ok, I did that.  Deprecated aliases added below in a separate block.

vstinner commented on Doc/c-api/hash.rst:
```diff
@@ -17,6 +17,29 @@ See also the :c:member:`PyTypeObject.tp_hash` member.
 
+.. c:macro:: PyHASH_MODULUS
+
+   The Mersenne prime ``P = 2**n -1``, used for numeric hash scheme.
```
I'm not sure that everybody knows about it. I suggest adding a link to Wikipedia:

```suggestion
   The `Mersenne prime <https://en.wikipedia.org/wiki/Mersenne_prime>`_ ``P = 2**n -1``, used for numeric hash scheme.
```

vstinner commented on Doc/c-api/hash.rst:
```diff
@@ -17,6 +17,29 @@ See also the :c:member:`PyTypeObject.tp_hash` member.
+
+.. c:macro:: PyHASH_BITS
+
+   The exponent ``n`` of ``P``.
```
```suggestion
   The exponent ``n`` of ``P`` in :c:macro:`PyHASH_MODULUS`.
```

skirpichev committed:
Apply suggestions from code review

Co-authored-by: Victor Stinner <vstinner@python.org>

skirpichev committed:
Apply suggestions from review: keep deprecated names as aliases

vstinner commented on Include/cpython/pyhash.h:
```diff
@@ -10,14 +10,19 @@
+#define PyHASH_INF 314159
+#define PyHASH_IMAG _PyHASH_MULTIPLIER
+
+#define _PyHASH_BITS PyHASH_BITS
```
```suggestion
// Aliases kept for backward compatibility with Python 3.12
#define _PyHASH_BITS PyHASH_BITS
```

skirpichev replied:
Done, but using ``/* style */`` for consistency with the rest of file.

vstinner reviewed (approved):
LGTM, I just suggest to add a comment.

@mdickinson: Would you mind to double check this PR?

skirpichev committed:
Add comment

mdickinson reviewed (approved):
Looks fine to me.
